src/libraries/Microsoft.PowerFx.Core/Functions/UserDefinedFunction.cs

```diff
@@ -0,0 +1,229 @@
+        }
+
+        /// <summary>
+        /// Perform sub-expression type checking and produce a return type for the function declaration, this is only applicable for UDFs.
+        /// </summary>
+        public static void CheckTypesOnDeclaration(CheckTypesContext context, IEnumerable<UDFArg> uDFArgs, IdentToken returnType, DType actualBodyReturnType, IErrorContainer errorContainer)
+        {
+            Contracts.AssertValue(context);
+            Contracts.AssertValue(uDFArgs);
+            Contracts.AssertValue(returnType);
+            Contracts.AssertValue(actualBodyReturnType);
+            Contracts.AssertValue(errorContainer);
+
+            CheckParameters(uDFArgs, errorContainer);
+            CheckReturnType(returnType, actualBodyReturnType, errorContainer);
+        }
+
+        private static void CheckReturnType(IdentToken returnType, DType actualBodyReturnType, IErrorContainer errorContainer)
+        {
+            var returnTypeFormulaType = returnType.GetFormulaType()._type;
+
+            if (returnTypeFormulaType.Kind.Equals(DType.Unknown.Kind))
+            {
+                errorContainer.EnsureError(DocumentErrorSeverity.Severe, returnType, TexlStrings.ErrUDF_UnknownType, returnType.Name);
+
+                // Anyhow the next error will be ignored, since it is added on the same token
+                return; 
+            }
+
+            if (!returnTypeFormulaType.Kind.Equals(actualBodyReturnType.Kind) || !returnTypeFormulaType.CoercesTo(returnTypeFormulaType))
+            {
+                errorContainer.EnsureError(DocumentErrorSeverity.Severe, returnType, TexlStrings.ErrUDF_ReturnTypeDoesNotMatch);
+            }
+        }
+
+        private static void CheckParameters(IEnumerable<UDFArg> args, IErrorContainer errorContainer)
+        {
+            foreach (var arg in args)
+            {
+                var argNameToken = arg.VarIdent;
+                var argTypeToken = arg.VarType;
+
+                if (argTypeToken.GetFormulaType()._type.Kind.Equals(DType.Unknown.Kind))
+                {
+                    errorContainer.EnsureError(DocumentErrorSeverity.Severe, argTypeToken, TexlStrings.ErrUDF_UnknownType, argTypeToken.Name);
+                }
+            }
+        }
+
+        public static StringGetter SG(string text)
+        {
+            return (string locale) => text;
+        }
+
+        public override IEnumerable<StringGetter[]> GetSignatures()
+        {
+            yield return new[] { SG("Arg 1") };
+        }
+
+        /// <summary>
+        /// NameResolver that combines global named resolver and params for user defined function.
+        /// </summary>
+        private class UserDefinitionsNameResolver : INameResolver
+        {
+            private readonly INameResolver _globalNameResolver;
+            private readonly INameResolver _parameterNamedResolver;
+
+            public static INameResolver Merge(INameResolver globalNameResolver, RecordType parametersRecordType)
```

Comment:
Where is the list of all visible UDFs? The function body may reference UDFs that were defined before or after this one.

Reply:
GlobalNameResolver will be able to resolve other UDFs, by the time bind is called, all the UDFs were already registered by the host.

Reply:
The host should not need to do any work to use this new API.

Reply:
Host is not doing any work to use the new API, Host calls the new API, that will parse all the named formulas and UDFs. It's now the host's responsibility to register the functions with it's own name resolver. During analysis we bind each UDF.